### PR TITLE
luminous: mds: underwater dentry check in CDir::_omap_fetched is racy

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/23157

### CODE REVIEW

**Comment by smithfarm** [comment]
retest this please

**Comment by smithfarm** [comment]
jenkins test docs

**Comment by smithfarm** [comment]
retest this please

**Comment by yuriw** [comment]
[wip-yuri3-testing-2018-05-07-1632-luminous](https://trello.com/c/vwdF3DiW/307-wip-yuri3-testing-2018-05-07-1632-luminous)

**Comment by yuriw** [comment]
@smithfarm suspect this broke the build 

/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.2.5-46-gbc29cfb/rpm/el7/BUILD/ceph-12.2.5-46-gbc29cfb/src/mds/CDir.cc:1972:57: error: no matching function for call to 'std::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)18u, char> >::basic_string(std::string&)'


see full log https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=huge/11440//consoleFull

File: src/mds/CDir.cc [hunk 1]
```diff
@@ -1937,35 +1962,16 @@ void CDir::_omap_fetched(bufferlist& hdrbl, map<string, bufferlist>& omap,
...
+    CDentry::linkage_t *dnl = dn->get_linkage();
+    if (dnl->is_primary() && dnl->get_inode()->state_test(CInode::STATE_REJOINUNDEF))
+      undef_inodes.push_back(dnl->get_inode());
 
-      if (dn->get_linkage()->is_primary()) {
-	assert(dn->get_linkage()->get_inode()->get_version() <= got_fnode.version);
-	dout(10) << "_fetched  had underwater inode " << *dn->get_linkage()->get_inode() << ", marking clean" << dendl;
-	dn->get_linkage()->get_inode()->mark_clean();
-      }
+    if (wanted_items.count(mempool::mds_co::string(dname)) > 0 || !complete) {
```
Comment (ukernel) [inline src/mds/CDir.cc]: wanted_items.count(mempool::mds_co::string(boost::string_view(dname))) > 0 ?

**Comment by ukernel** [comment]
The updated code looks exact the same as the code in wip-luminous-integration-20180530

http://pulpito.ceph.com/?branch=wip-luminous-integration-20180530

**Pull request merged**